### ISSUE TITLE
Intermittent crashes with NSUrlSessionHandler

### ISSUE THREAD

**Issue by filipnavara** [xamarin/xamarin-macios#13921]
Unfortunately I don't have reliable repro yet but it consistently happens in our application when connecting to a specific endpoint. The crash doesn't happen for every request, only after the application runs for a while, which makes it difficult to capture.

```
Exception thrown: 'System.NullReferenceException' in System.Private.CoreLib.dll
Exception thrown: 'System.NullReferenceException' in System.Private.CoreLib.dll
Exception thrown: 'System.NullReferenceException' in System.Private.CoreLib.dll
2022-01-27 11:26:02.272 eM Client[19301:115633] *** Terminating app due to uncaught exception 'System.ArgumentException', reason: 'Object of type 'Foundation.NSUrlSessionTask' cannot be converted to type 'Foundation.NSUrlSessionDataTask'.'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001855d01cc __exceptionPreprocess + 240
	1   libobjc.A.dylib                     0x00000001853217b8 objc_exception_throw + 60
	2   eM Client                           0x00000001049358fc xamarin_process_managed_exception + 796
	3   eM Client                           0x00000001049bfa9c _ZL32native_to_managed_trampoline_185P11objc_objectP13objc_selectorPP11_MonoObjectS0_S0_S0_S0_j + 884
	4   eM Client                           0x00000001049bf71c -[Foundation_NSUrlSessionHandler_NSUrlSessionHandlerDelegate URLSession:dataTask:willCacheResponse:completionHandler:] + 76
	5   CFNetwork                           0x000000018a1c3450 _CFHostIsDomainTopLevelForCertificatePolicy + 29804
	6   libdispatch.dylib                   0x00000001852c4e60 _dispatch_call_block_and_release + 32
	7   libdispatch.dylib                   0x00000001852c6bac _dispatch_client_callout + 20
	8   libdispatch.dylib                   0x00000001852ce330 _dispatch_lane_serial_drain + 672
	9   libdispatch.dylib                   0x00000001852ceed8 _dispatch_lane_invoke + 444
	10  libdispatch.dylib                   0x00000001852d9708 _dispatch_workloop_worker_thread + 656
	11  libsystem_pthread.dylib             0x0000000185481304 _pthread_wqthread + 288
	12  libsystem_pthread.dylib             0x0000000185480018 start_wqthread + 8
)
libc++abi: terminating with uncaught exception of type NSException
```

Environment: .NET 6.0.1, macOS Monterey 12.2 RC

**Comment by rolfbjarne** [comment]
Ugh, this is annoying.

The method in question has a `NSUrlSessionDataTask` parameter, and sometimes Apple gives us a `NSUrlSessionTask` (or custom subclass thereof) that just quacks like a `NSUrlSessionDataTask`, while not actually being a `NSUrlSessionDataTask` (or a subclass).

We already had to come up with a workaround this for scenario (but the same types) some time ago.

My guess is that it happens when download tasks are moved to the background.

**Comment by filipnavara** [comment]
> My guess is that it happens when download tasks are moved to the background.

Looks like pretty reasonable guess. I had the app running in foreground for 20 minutes and could not reproduce it. Now I had a browser in the foreground for a moment and it crashed while the app was in the background.

**Comment by filipnavara** [comment]
Duplicate of https://github.com/xamarin/xamarin-macios/issues/13531 btw.

**Comment by filipnavara** [comment]
This is really difficult to reproduce in a smaller app. I can reliably reproduce on the full app though and it changes behavior based on debugger exception checkpoints (VS Code debugger). One thing I noticed that there's usually couple of `NullReferenceException`/`ObjectDisposedException`s preceding the final crash. In the VS4Mac debugger on a sample app I managed to catch one of them:

```
System.ObjectDisposedException: "Cannot access a disposed object.
Object name: 'Foundation.NSMutableString'."
  at ObjCRuntime.ThrowHelper.ThrowObjectDisposedException(Object o)
   at Foundation.NSDictionary.ObjectForKey(NSObject key)
   at Foundation.NSDictionary.<GetEnumerator>d__64.MoveNext()
   at Foundation.NSUrlSessionHandler.NSUrlSessionHandlerDelegate.DidReceiveResponse(NSUrlSession session, NSUrlSessionDataTask dataTask, NSUrlResponse response, Action`1 completionHandler)
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Foundation.NSUrlSessionHandler.<SendAsync>d__57.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<<SendAsync>g__Core|83_0>d.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at nsurlsession.AppDelegate.<RunHttpRequest>d__2.MoveNext() in /Users/filipnavara/Projects/nsurlsession/AppDelegate.cs:54
```

The app basically does couple of GETs/POSTs on a background thread in a loop, interleaved with `GC.Collect` calls.

**Comment by filipnavara** [comment]
I can reproduce the crash by creating an empty app with `dotnet new macos`, adding the following code:
```csharp
    static async void RunHttpRequest()
    {
        using var handler = new NSUrlSessionHandler();
        handler.CookieContainer = new CookieContainer();
        handler.Credentials = new NetworkCredential("TESTUSER", "grundlE!12345", "emclientntlm");

        while (true)
        {
            try
            {
                using var client = new HttpClient(handler);
                var message = new HttpRequestMessage(HttpMethod.Get, "http://emclientntlm.westus.cloudapp.azure.com/");
                var result = await client.SendAsync(message);
                Console.WriteLine(result);
            }
            catch (HttpRequestException e)
            {
                Console.WriteLine(e);
            }
            catch (TaskCanceledException e)
            {
                Console.WriteLine(e);
            }

            await Task.Delay(100);
        }
    }
```
and then calling `RunHttpRequest();` from `DidFinishLaunching`. It eventually crashes with either the `ArgumentException` or `ObjectDisposedException`. The issue is that it takes anything from a second to an hour before it happens.

(The URL and credentials are on a throw-away virtual server so it's fine to hammer it or to share the credentials.)

**Comment by filipnavara** [comment]
Got one more exception (same repro code):
```
Exception has occurred: CLR/System.ArgumentNullException
An exception of type 'System.ArgumentNullException' occurred in System.Private.CoreLib.dll but was not handled in user code: 'Value cannot be null.'
   at System.IO.UnmanagedMemoryStream.Initialize(Byte* pointer, Int64 length, Int64 capacity, FileAccess access)
   at Foundation.NSData.AsStream()
   at Foundation.NSUrlSessionHandler.NSUrlSessionDataTaskStream.<ReadAsync>d__14.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.IO.Stream.<<CopyToAsync>g__Core|29_0>d.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpContent.<LoadIntoBufferAsyncCore>d__63.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<<SendAsync>g__Core|83_0>d.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at nsurlsession.AppDelegate.<RunWikiHttpRequest>d__3.MoveNext() in /Users/filipnavara/Projects/nsurlsession/AppDelegate.cs:line 42
```

**Comment by filipnavara** [comment]
And finally managed to get the `NullReferenceException` too:

```csharp
					if (session.Configuration.HttpCookieStorage != null) {
						var cookies = session.Configuration.HttpCookieStorage.CookiesForUrl (response.Url);
						UpdateManagedCookieContainer (response.Url, cookies);
```

^^^ `cookies` was `null` in here.

Updated repro which usually crashes within minute: [Archive.zip](https://github.com/xamarin/xamarin-macios/files/7986770/Archive.zip). I now run the same request in four parallel sessions. Additionally I copied `NSUrlSessionHandler` into the project to get more reliable stack traces and debugging information.

**Comment by rolfbjarne** [comment]
I wonder if this might be some of the reasons for the problems people report in #9132.

**Comment by filipnavara** [comment]
I think I found the root cause.

There's global dictionary of `NSObject`s that are tracked with `GCHandle`. It uses the `WeakTrackResurrection` type. When the finalizer runs it calls `NSObject_Disposer.Add(this)`. That resurrects the object so for a brief period there is a resurrected & disposed object. `Runtime.TryGetNSObject` would check the `GCHandle`, find it and return it.

Since the object is not subscribed for GC tracking (the coreclr-bridge.m logic) it also never has the `InFinalizerQueue` bit set.

**Comment by filipnavara** [comment]
Minimal repro of the GC issue (deterministic and single threaded):

```
public class Program
{
	[System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
	public static IntPtr CreateNativeObject()
	{
		return new NSString("foo").Handle;
	}

	public static void PumpLoop()
	{
		using (new NSAutoreleasePool())
		{
			NSEvent? evnt;
			while ((evnt = NSApplication.SharedApplication.NextEvent((NSEventMask)NSEventMask.AnyEvent, NSDate.DistantPast, NSRunLoopMode.Default, true)) != null)
			{
				NSApplication.SharedApplication.SendEvent(evnt);
			}
		}
	}

	public static void NSObjectFinalizeTest()
	{
		// Pump run loop to ensure that NSObject_Disposer has processed everything
		PumpLoop();

		// Create a single object and immediately lose a reference to it, save the handle
		// to immitate getting the same handle from a native API at later point in time
		IntPtr nativeHandle = CreateNativeObject();

		// Ensure that the GC is run and the one new object gets collected
		GC.Collect();
		GC.WaitForPendingFinalizers();

		// The object should now be resurrected through `NSObject_Disposer.Add(this)`
		// now and so `GCHandle` should return it's reference:
		var o = ObjCRuntime.Runtime.TryGetNSObject(nativeHandle);
		Console.WriteLine("Handle (1): " + o.Handle);

		// Pump the run loop and thus drain the NSObject_Disposer
		PumpLoop();

		// The object is invalid now
		Console.WriteLine("Handle (2): " + o.Handle);
	}

	public static void Main()
	{
		NSApplication.Init();
		NSObjectFinalizeTest();
	}
}
```

**Comment by filipnavara** [comment]
FWIW this issue actually caused quite a few crashes in our app. We only managed to attribute them later to the same GC fault. I am posting it here in case someone happens to stumble upon it or Googles it.

One place where we hit it is with global shared objects. For example, `var x = NSSpellChecker.SharedSpellChecker; ... foreach (var l in x.AvailableLanguages) { ... }` would randomly crash with `NullReferenceException`. The underlying cause was the same. During the course of the execution `NSSpellChecker.SharedSpellChecker` would return an instance that was previously finalized and by the time we call `AvailableLanguages` on it the handle was already cleared through the UI loop.

**Comment by filipnavara** [comment]
Unfortunately this still happens, so the issue is not resolved completely. The root cause is still connected to the GC interaction with the object cache but it appears to be used incorrectly somewhere in the callbacks from the native code. I noticed that there's couple of places that call `GetNSObject(..., evenInFinalizerQueue: true, ...)` which are called from the static registrar and look suspicious. My understanding of the area is not good enough to decide whether it is intentional or an unintended race condition.

```
2022-03-30 19:49:32.344 eM Client[98648:3141146] *** Terminating app due to uncaught exception 'System.ArgumentException', reason: 'Object of type 'Foundation.NSUrlSessionTask' cannot be converted to type 'Foundation.NSUrlSessionDataTask'. (System.ArgumentException)
   at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast)
   at System.Reflection.MethodBase.CheckArguments(StackAllocedArguments& stackArgs, ReadOnlySpan`1 parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at ObjCRuntime.Runtime.InvokeMethod(MethodBase method, Object instance, IntPtr native_parameters)
   at ObjCRuntime.Runtime.InvokeMethod(MonoObject* methodobj, MonoObject* instanceobj, IntPtr native_parameters)
   at ObjCRuntime.Runtime.bridge_runtime_invoke_method(MonoObject* method, MonoObject* instance, IntPtr parameters, IntPtr& exception_gchandle)
'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001a1a36b08 __exceptionPreprocess + 240
	1   libobjc.A.dylib                     0x00000001a1781e14 objc_exception_throw + 60
	2   eM Client                           0x0000000100d8b464 xamarin_process_managed_exception + 752
	3   eM Client                           0x0000000100e12408 _ZL32native_to_managed_trampoline_183P11objc_objectP13objc_selectorPP11_MonoObjectS0_S0_S0_S0_j + 884
	4   eM Client                           0x0000000100e12088 -[Foundation_NSUrlSessionHandler_NSUrlSessionHandlerDelegate URLSession:dataTask:willCacheResponse:completionHandler:] + 76
	5   CFNetwork                           0x00000001a66000e8 _CFHostIsDomainTopLevelForCertificatePolicy + 28772
	6   libdispatch.dylib                   0x00000001a1724604 _dispatch_call_block_and_release + 32
	7   libdispatch.dylib                   0x00000001a17261c8 _dispatch_client_callout + 20
	8   libdispatch.dylib                   0x00000001a172d8a8 _dispatch_lane_serial_drain + 668
	9   libdispatch.dylib                   0x00000001a172e438 _dispatch_lane_invoke + 444
	10  libdispatch.dylib                   0x00000001a1738c98 _dispatch_workloop_worker_thread + 648
	11  libsystem_pthread.dylib             0x00000001a18e6360 _pthread_wqthread + 288
	12  libsystem_pthread.dylib             0x00000001a18e5080 start_wqthread + 8
)
```

**Comment by filipnavara** [comment]
Repro:
[Archive.zip](https://github.com/xamarin/xamarin-macios/files/8383411/Archive.zip)

Usage:
- `dotnet build`
- `./bin/Debug/net6.0-macos/gctest1.app/Contents/MacOS/gctest1`

Main code is in `ViewController.cs`, rest is empty app template. On start it spawns one task that runs GC every 3ms and five tasks that hammer https://google.com with simple GET request every 10ms. In few seconds it will crash with one of the errors above. On osx-x64 `ArgumentException` is more common, on osx-arm64 `ObjectDisposedException` is more common, but in the end both can happen on any platform.

**Comment by rolfbjarne** [comment]
Let's continue in #13531, to avoid confusion.
